base_layer/core/src/base_node/rpc/service.rs

```diff
@@ -0,0 +1,194 @@
+                            .await
+                            .map_err(RpcStatus::log_internal_error(LOG_TARGET))?
+                        {
+                            None => TxSubmissionResponse {
+                                accepted: false,
+                                rejection_reason: TxSubmissionRejectionReason::DoubleSpend.into(),
+                            },
+                            Some(_) => TxSubmissionResponse {
+                                accepted: false,
+                                rejection_reason: TxSubmissionRejectionReason::AlreadyMined.into(),
+                            },
+                        }
+                    },
+                }
+            },
+        };
+        Ok(Response::new(response))
+    }
+
+    async fn transaction_query(
+        &self,
+        request: Request<SignatureProto>,
+    ) -> Result<Response<TxQueryResponse>, RpcStatus>
+    {
+        let message = request.into_message();
+        let signature = Signature::try_from(message).map_err(|_| RpcStatus::bad_request("Signature was invalid"))?;
+
+        let db = self.db();
+        match db
+            .fetch_kernel_by_excess_sig(signature.clone())
+            .await
+            .map_err(RpcStatus::log_internal_error(LOG_TARGET))?
+        {
+            None => (),
+            Some((_, block_hash)) => {
+                match db
+                    .fetch_header_by_block_hash(block_hash.clone())
+                    .await
+                    .map_err(RpcStatus::log_internal_error(LOG_TARGET))?
+                {
+                    None => (),
+                    Some(header) => {
+                        let chain_meta_data = db
+                            .get_chain_metadata()
+                            .await
+                            .map_err(RpcStatus::log_internal_error(LOG_TARGET))?;
+                        let confirmations = chain_meta_data.height_of_longest_chain() - header.height;
+                        let response = TxQueryResponse {
+                            location: TxLocation::Mined as i32,
+                            block_hash: Some(block_hash),
+                            confirmations,
+                        };
+                        return Ok(Response::new(response));
+                    },
+                }
+            },
+        };
+
+        // If not in a block then check the mempool
+        let mut mempool = self.mempool();
+        let mempool_response = match mempool
+            .get_tx_state_by_excess_sig(signature.clone())
+            .await
+            .map_err(RpcStatus::log_internal_error(LOG_TARGET))?
+        {
+            TxStorageResponse::UnconfirmedPool => TxQueryResponse {
+                location: TxLocation::InMempool as i32,
+                block_hash: None,
+                confirmations: 0,
+            },
+            TxStorageResponse::ReorgPool |
+            TxStorageResponse::NotStoredOrphan |
```

Comment:
Is it possible for the mempool to ever return these? Surely the mempool will not know the reason something is not stored, because it has not stored it

Reply:
The validator tells it what the issue was, that's how it used to decide which pool to put it in.

Reply:
oh on the query you are right its not possible

Reply:
I agree on this change and I got started but turns out that the same struct is used for the service handle interface that is used by the proto of the mempool service so making an update to way the handle returns involves updating everything in the repo that touches the proto too which is extensive. I will make an issue to fix this but I think it should be its own PR.